### PR TITLE
Add debug widget into 3D view

### PR DESCRIPTION
**Pull request by Withalion**
## Description

The new debug side panel moves all relevant debug settings into one place, from multiple into one easily accessible side 
panel. Furthermore the change implemented in #58947 was missing any option to toggle this on/off. 

Features in this PR:
- New checkbox for stopping scene updates
- Options moved from advanced sub-category
- Camera info moved from navigation side panel and changed the way it's shown (changing values through the new spin boxes in next PR)
- Debug depth map view default position moved from top right corner into bottom left, so it won't collide with the axis view
- The size of toolbar icons is responsive to the docked state of window  

Standalone window:
Screenshot_20241118_224841

Docked window:
Screenshot_20241118_224929

### CODE REVIEW

File: src/app/3d/qgs3dmapcanvaswidget.cpp [hunk 1]
```diff
@@ -305,6 +308,19 @@ Qgs3DMapCanvasWidget::Qgs3DMapCanvasWidget( const QString &name, bool isDocked )
 
   onTotalPendingJobsCountChanged();
 
+  QAction *toggleDebugPanel = toolBar->addAction(
```
Comment (nyalldawson) [inline src/app/3d/qgs3dmapcanvaswidget.cpp:311]: I don't think we should make this visible by default. It's not something the average user ever needs, and adding it to the toolbar just contributes to UI clutter. Instead of a toolbar action, how about some hidden key shortcut to toggle the pane?

Comment (uclaros) [reply src/app/3d/qgs3dmapcanvaswidget.cpp:311]: I agree, but I'd rather have a `Show debug panel` checkbox in advanced settings page instead of completely hiding it!

Comment (Withalion) [reply src/app/3d/qgs3dmapcanvaswidget.cpp:311]: I like the @uclaros approach of having a checkbox + I can add a keyboard shortcut

Comment (Withalion) [reply src/app/3d/qgs3dmapcanvaswidget.cpp]: I moved the setting into advanced sub-category of options and added a Ctrl+Shift+d shortcut for the panel

**Comment by Withalion** [comment]
Regarding the checkboxes, yes totally agree however that's the theme issue. I'm using the default theme in dark mode, other themes don't have this issue. Of course the issue is in the whole app not just in the debug panel

File: src/3d/qgs3dmapsettings.cpp [hunk 2]
```diff
@@ -70,6 +70,7 @@ Qgs3DMapSettings::Qgs3DMapSettings( const Qgs3DMapSettings &other )
   , mShowCameraRotationCenter( other.mShowCameraRotationCenter )
   , mShowLightSources( other.mShowLightSources )
   , mShowLabels( other.mShowLabels )
+  , mStopUpdates( other.mStopUpdates )
```
Comment (wonder-sk) [inline src/3d/qgs3dmapsettings.cpp:73]: probably also `mShowDebugPanel` should be copied

File: src/3d/qgs3dmapsettings.cpp [hunk 3]
```diff
@@ -313,6 +314,7 @@ void Qgs3DMapSettings::readXml( const QDomElement &elem, const QgsReadWriteConte
   mShowCameraRotationCenter = elemDebug.attribute( QStringLiteral( "camera-rotation-center" ), QStringLiteral( "0" ) ).toInt();
   mShowLightSources = elemDebug.attribute( QStringLiteral( "show-light-sources" ), QStringLiteral( "0" ) ).toInt();
   mIsFpsCounterEnabled = elemDebug.attribute( QStringLiteral( "show-fps-counter" ), QStringLiteral( "0" ) ).toInt();
+  mStopUpdates = elemDebug.attribute( QStringLiteral( "stop-updates" ), QStringLiteral( "0" ) ).toInt();
```
Comment (wonder-sk) [inline src/3d/qgs3dmapsettings.cpp:317]: probably also `mShowDebugPanel` should be read from / written to xml...

File: src/3d/qgs3dmapsettings.h [hunk 4]
```diff
@@ -431,6 +431,14 @@ class _3D_EXPORT Qgs3DMapSettings : public QObject, public QgsTemporalRangeObjec
     //! Returns the eye dome lighting distance value (contributes to the contrast of the image)
     int eyeDomeLightingDistance() const;
 
+    /**
+    * Sets whether to keep updating scene on zooming or keep the same amount of points visible
+    * \since QGIS 3.42
+    */
+    void setStopUpdates( bool enabled );
+    //! Returns whether to keep updating the scene on zoom
+    bool stopUpdates() const;
```
Comment (wonder-sk) [inline src/3d/qgs3dmapsettings.h:440]: ideally this method should be also marked with `\since QGIS 3.42`

File: src/3d/qgs3dmapsettings.h [hunk 5]
```diff
@@ -431,6 +431,14 @@ class _3D_EXPORT Qgs3DMapSettings : public QObject, public QgsTemporalRangeObjec
     //! Returns the eye dome lighting distance value (contributes to the contrast of the image)
     int eyeDomeLightingDistance() const;
 
+    /**
+    * Sets whether to keep updating scene on zooming or keep the same amount of points visible
```
Comment (wonder-sk) [inline src/3d/qgs3dmapsettings.h:435]: this "stop updates" does not apply only to points, but pretty much all 3D entities - terrain, vector layers, tiled scene layers etc. So best to rephrase this to something like "Sets whether scene updates upon camera movement should be enabled", maybe even add a note that by default scene always gets updated and this option is good for debugging purposes.

File: src/app/3d/qgs3ddebugwidget.cpp [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+ ***************************************************************************/
+
+#include "qgis.h"
+
+#include "moc_qgs3ddebugwidget.cpp"
+#include "qgs3ddebugwidget.h"
+#include "qgs3dmapcanvas.h"
+#include "qgscameracontroller.h"
+
+Qgs3DDebugWidget::Qgs3DDebugWidget( Qgs3DMapCanvas *canvas, QWidget *parent ) : QWidget( parent ), m3DMapCanvas( canvas )
```
Comment (wonder-sk) [inline src/app/3d/qgs3ddebugwidget.cpp:23]: it is a habit in qgis code to put the initializer list on the next line, and have each member initialized on a separate line for better readability

File: src/app/3d/qgs3ddebugwidget.cpp [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+  mCameraY->setRange( std::numeric_limits<float>::lowest(), std::numeric_limits<float>::max() );
+  mCameraZ->setRange( std::numeric_limits<float>::lowest(), std::numeric_limits<float>::max() );
+  mLookingX->setRange( std::numeric_limits<float>::lowest(), std::numeric_limits<float>::max() );
+  mLookingY->setRange( std::numeric_limits<float>::lowest(), std::numeric_limits<float>::max() );
+  mLookingZ->setRange( std::numeric_limits<float>::lowest(), std::numeric_limits<float>::max() );
+}
+
+/**
+ * Sets up the interactive elements with values from Qgs3DMapSettings
+ */
```
Comment (wonder-sk) [inline src/app/3d/qgs3ddebugwidget.cpp:64]: Common practice in QGIS code is to have function docstrings in the header... (also for updateFromCamera())

File: src/app/3d/qgs3ddebugwidget.h [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+ ***************************************************************************/
+
+#ifndef QGS3DDEBUGWIDGET_H
+#define QGS3DDEBUGWIDGET_H
+
+class QStandardItemModel;
+class Qgs3DMapCanvas;
+
+#include <ui_3ddebugwidget.h>
+#include "qgs3dmapsettings.h"
```
Comment (wonder-sk) [inline src/app/3d/qgs3ddebugwidget.h:23]: we don't need to include qgs3dmapsettings.h here, only in the cpp - here we can have just forward declaration

File: src/app/3d/qgs3dmapcanvaswidget.cpp [hunk 9]
```diff
@@ -288,12 +289,18 @@ Qgs3DMapCanvasWidget::Qgs3DMapCanvasWidget( const QString &name, bool isDocked )
...
   mNavigationWidget->setSizePolicy( QSizePolicy::Fixed, QSizePolicy::Expanding );
+  mDebugWidget = new Qgs3DDebugWidget( mCanvas );
 
   QHBoxLayout *hLayout = new QHBoxLayout;
   hLayout->setContentsMargins( 0, 0, 0, 0 );
   hLayout->addWidget( mContainer );
   hLayout->addWidget( mNavigationWidget );
+  hLayout->addWidget( mDebugWidget );
 
+  auto *debugPanelShortCut = new QShortcut( QKeySequence( tr( "Ctrl+Shift+d" ) ), this );
```
Comment (wonder-sk) [inline src/app/3d/qgs3dmapcanvaswidget.cpp:300]: It is a habit in QGIS code to only use `auto` for complex types (e.g. iterators) - for "simple" types it is preferred to write the type for code clarity (arguably here's it is clear, but in other places it may not be as clear, e.g. when assigning returned value from a function)

File: src/app/3d/qgs3ddebugwidget.cpp [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+  setupUi( this );
+
+  // set up the fixed width of debug widget
+  mCameraInfoGroupBox->setMinimumWidth( mCameraInfoGroupBox->sizeHint().width() );
+  mCameraInfoGroupBox->adjustSize();
+  scrollAreaWidgetContents->setMinimumWidth( scrollAreaWidgetContents->sizeHint().width() );
+  scrollAreaWidgetContents->adjustSize();
+  scrollArea->setMinimumWidth( scrollArea->sizeHint().width() );
+  scrollArea->adjustSize();
+  this->adjustSize();
```
Comment (wonder-sk) [inline src/app/3d/qgs3ddebugwidget.cpp:35]:
```suggestion
  adjustSize();
```
(best to avoid `this->` unless it is really needed)

**Review by wonder-sk — COMMENTED** [review]
Looking good - finally we're on the way to proper developer tools for 3D scenes :slightly_smiling_face:

File: src/app/3d/qgs3dmapcanvaswidget.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/app/3d/qgs3dmapcanvaswidget.cpp]: @Withalion 
> Ctrl+Shift+d 

That's already used for opening Data Source Manager. How about F12? That's used for the dev tools panel in the main window, so I think it'd be nice to use that here too. (I realise that if the 3d view is docked then F12 would only work if the 3d view is focused, but that same behavior would also be present with Ctrl+Shift+D due to the conflict with the Data Source Manager shortcut)

File: src/app/3d/qgs3ddebugwidget.h (hunk 8 above)
Comment (Withalion) [reply src/app/3d/qgs3ddebugwidget.h:23]: compiler doesn't agree on this :( so it will have to stay

File: src/app/3d/qgs3dmapcanvaswidget.cpp (hunk 1 above)
Comment (Withalion) [reply src/app/3d/qgs3dmapcanvaswidget.cpp]: Okay that makes sense. I guess having an overlap in shortcuts is unavoidable, but it's definitely better UX on F12.

Comment (Withalion) [reply src/app/3d/qgs3dmapcanvaswidget.cpp]: @nyalldawson after testing the change, F12 on docked windows doesn't work at all. For the previous combination I at least got a window about shortcut ambiguity, but on F12 I don't get anything. It just doesn't work. In stand alone window it works.

**Review by wonder-sk — APPROVED** [review]
